rllib/core/rl_module/rl_module.py

```diff
@@ -0,0 +1,194 @@
+import abc
+from typing import Mapping, Any
+
+from ray.rllib.utils.annotations import (
+    ExperimentalAPI,
+    OverrideToImplementCustomLogic_CallToSuperRecommended,
+)
+
+from ray.rllib.models.specs.specs_dict import ModelSpec, check_specs
+from ray.rllib.models.distributions import Distribution
+from ray.rllib.utils.nested_dict import NestedDict
+from ray.rllib.utils.typing import SampleBatchType
+
+
+@ExperimentalAPI
+class RLModule(abc.ABC):
+    """Base class for RLlib modules.
+
+    Here is the pseudo code for how the forward methods are called:
+
+    # During Training (acting in env from each rollout worker)
+    ----------------------------------------------------------
+    .. code-block:: python
+
+        module: RLModule = ...
+        obs = env.reset()
+        while not done:
+            fwd_outputs = module.forward_exploration({"obs": obs})
+            # this can be deterministic or stochastic exploration
+            action = fwd_outputs["action_dist"].sample()
+            next_obs, reward, done, info = env.step(action)
+            buffer.add(obs, action, next_obs, reward, done, info)
+            next_obs = obs
+
+    # During Training (learning the policy)
+    ----------------------------------------------------------
+    .. code-block:: python
+        module: RLModule = ...
+        fwd_ins = buffer.sample()
+        fwd_outputs = module.forward_train(fwd_ins)
+        loss = compute_loss(fwd_outputs, fwd_ins)
+        update_params(module, loss)
+
+    # During Inference (acting in env during evaluation)
+    ----------------------------------------------------------
+    .. code-block:: python
+        module: RLModule = ...
+        obs = env.reset()
+        while not done:
+            fwd_outputs = module.forward_inference({"obs": obs})
+            # this can be deterministic or stochastic evaluation
+            action = fwd_outputs["action_dist"].sample()
+            next_obs, reward, done, info = env.step(action)
+            next_obs = obs
+
+    Args:
+        config: The config object for the module.
+
+    Abstract Methods:
+        forward_train: Forward pass during training.
+        forward_exploration: Forward pass during training for exploration.
+        forward_inference: Forward pass during inference.
```

Comment:
when did we change ``forward()`` and ``forward(exploration=True)`` to ``forward_inference()`` and ``forward_exploration()``?
it's kind of verbose.

Reply:
They may have different input / output requirements that is cannot be controlled with the kwarg. I also found it more useful to be more explicit about these potentially different behaviors. You can think of this sort of as similar to `training_step`, `validation_step` and `test_step` methods in [pytorch_lightning.](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html)